## Importing Libraries

In [1]:
import pandas as pd
import optuna

from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, StackingRegressor
from xgboost import XGBRegressor
from catboost import CatBoostRegressor, Pool, cv
from lightgbm import LGBMRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error, mean_absolute_error, root_mean_squared_error, r2_score

## Configurations

In [2]:
# path of training dataset
data_path = 'data/processed/train_processed.csv'

FOLDS=5 # no. of folds for cross validation
SPLIT=0.2 # splitting size
SEED=42 # seed for reproducibility

In [3]:
# loading the training dataset
df = pd.read_csv(data_path)

In [4]:
# separating dependent and independent variables
X = df.drop('MedHouseVal', axis=1)
y = df['MedHouseVal']

In [5]:
# splitting the dataset into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=SPLIT, random_state=SEED)

## Model Training

In [6]:
# models for training
models = {
    'LinearRegression': LinearRegression(),
    'Lasso Regression': Lasso(),
    'Ridge Regression': Ridge(),
    'DecisionTreeRegressor': DecisionTreeRegressor(),
    'RandomForestRegressor': RandomForestRegressor(),
    'GradientBoostingRegressor': GradientBoostingRegressor(),
    'XGBRegressor': XGBRegressor(),
    'CatBoostRegressor': CatBoostRegressor(verbose=False),
    'LGBMRegressor': LGBMRegressor(verbose=-1),
    'KNeighborsRegressor': KNeighborsRegressor(),
    'Support Vector Regressor': SVR(),
    
}

In [7]:
# comparing models
results = []
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_val)

    results.append({
        'Model': name,
        'MSE': mean_squared_error(y_val, y_pred),
        'MAE': mean_absolute_error(y_val, y_pred),
        'RMSE': root_mean_squared_error(y_val, y_pred),
        'R2 Score': r2_score(y_val, y_pred)
    })

    results_df = pd.DataFrame(results).sort_values('R2 Score', ascending=False)

In [8]:
results_df

Model       MSE       MAE      RMSE  R2 Score
7           CatBoostRegressor  0.044521  0.158761  0.210999  0.649946
6                XGBRegressor  0.045057  0.159137  0.212266  0.645731
8               LGBMRegressor  0.045186  0.160274  0.212571  0.644712
4       RandomForestRegressor  0.047826  0.162961  0.218692  0.623955
5   GradientBoostingRegressor  0.054454  0.179304  0.233354  0.571841
9         KNeighborsRegressor  0.064928  0.192371  0.254810  0.489487
10   Support Vector Regressor  0.074108  0.208277  0.272228  0.417308
3       DecisionTreeRegressor  0.078355  0.207861  0.279920  0.383911
2            Ridge Regression  0.088286  0.234646  0.297130  0.305830
0            LinearRegression  0.088287  0.234644  0.297132  0.305818
1            Lasso Regression  0.127191  0.290428  0.356638 -0.000069

## Tuning Hyper-Parameters using Optuna

### 1. CatBoostRegressor

In [13]:
# objective function for tuning CatBoostRegressor
def objective(trail):
    params = {
        'iterations': trail.suggest_int('iterations', 500, 1500),
        'learning_rate': trail.suggest_float('learning_rate', 0.01, 0.1),
        'depth': trail.suggest_int('depth', 1, 10),
        'l2_leaf_reg': trail.suggest_int('l2_leaf_reg', 1, 10),
        'border_count': trail.suggest_int('border_count',10, 150),
        'random_strength': trail.suggest_float('random_strength', 0.1, 1.0),
        'bagging_temperature': trail.suggest_float('bagging_temperature', 0, 1.0),

        'loss_function': 'RMSE',
        'eval_metric': 'R2',
        'random_seed': SEED,
        'verbose': False
    }

    cv_results = cv(
        pool=Pool(X_train, y_train),
        params=params,
        fold_count=FOLDS,
        shuffle=True,
        stratified=False,
        verbose=False
    )

    best_r2 = max(cv_results['test-R2-mean'])

    return best_r2

In [14]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-14 19:11:36,415] A new study created in memory with name: no-name-ac7f52f2-0d62-4516-8967-30e9a52cc7ee


Training on fold [0/5]

bestTest = 0.6349491216
bestIteration = 1364

Training on fold [1/5]

bestTest = 0.6269462284
bestIteration = 1364

Training on fold [2/5]

bestTest = 0.6308059756
bestIteration = 1364

Training on fold [3/5]

bestTest = 0.6314658907
bestIteration = 1362

Training on fold [4/5]


[I 2026-02-14 19:11:54,108] Trial 0 finished with value: 0.6306903474315056 and parameters: {'iterations': 1365, 'learning_rate': 0.04545321069179818, 'depth': 3, 'l2_leaf_reg': 3, 'border_count': 101, 'random_strength': 0.35413665127826044, 'bagging_temperature': 0.9064442446934077}. Best is trial 0 with value: 0.6306903474315056.



bestTest = 0.6293122791
bestIteration = 1363

Training on fold [0/5]

bestTest = 0.3927428083
bestIteration = 606

Training on fold [1/5]

bestTest = 0.397548966
bestIteration = 606

Training on fold [2/5]

bestTest = 0.393170206
bestIteration = 606

Training on fold [3/5]

bestTest = 0.3999037335
bestIteration = 606

Training on fold [4/5]


[I 2026-02-14 19:12:00,543] Trial 1 finished with value: 0.3957049375107343 and parameters: {'iterations': 607, 'learning_rate': 0.03758490624037313, 'depth': 1, 'l2_leaf_reg': 1, 'border_count': 112, 'random_strength': 0.32086249314172083, 'bagging_temperature': 0.6526851739720636}. Best is trial 0 with value: 0.6306903474315056.



bestTest = 0.3951589737
bestIteration = 606

Training on fold [0/5]

bestTest = 0.6609239343
bestIteration = 1034

Training on fold [1/5]

bestTest = 0.6484618008
bestIteration = 821

Training on fold [2/5]

bestTest = 0.6536983978
bestIteration = 718

Training on fold [3/5]

bestTest = 0.6532462268
bestIteration = 842

Training on fold [4/5]


[I 2026-02-14 19:12:41,665] Trial 2 finished with value: 0.652972744323651 and parameters: {'iterations': 1407, 'learning_rate': 0.07558872380549825, 'depth': 9, 'l2_leaf_reg': 2, 'border_count': 91, 'random_strength': 0.8928441070693371, 'bagging_temperature': 0.8889115800572993}. Best is trial 2 with value: 0.652972744323651.



bestTest = 0.6496253511
bestIteration = 910

Training on fold [0/5]

bestTest = 0.6526132742
bestIteration = 1055

Training on fold [1/5]

bestTest = 0.6401239122
bestIteration = 1055

Training on fold [2/5]

bestTest = 0.6430899083
bestIteration = 1055

Training on fold [3/5]

bestTest = 0.6420398319
bestIteration = 1055

Training on fold [4/5]


[I 2026-02-14 19:13:33,059] Trial 3 finished with value: 0.644327990110997 and parameters: {'iterations': 1056, 'learning_rate': 0.014982865211429337, 'depth': 10, 'l2_leaf_reg': 3, 'border_count': 117, 'random_strength': 0.9818337066607737, 'bagging_temperature': 0.9419658904639207}. Best is trial 2 with value: 0.652972744323651.



bestTest = 0.643773024
bestIteration = 1055

Training on fold [0/5]

bestTest = 0.4462872216
bestIteration = 293

Training on fold [1/5]

bestTest = 0.4344845741
bestIteration = 348

Training on fold [2/5]

bestTest = 0.4724946581
bestIteration = 348

Training on fold [3/5]

bestTest = 0.5216376828
bestIteration = 258

Training on fold [4/5]


[I 2026-02-14 19:13:55,831] Trial 4 finished with value: 0.46473710550254177 and parameters: {'iterations': 1361, 'learning_rate': 0.07288342034253585, 'depth': 7, 'l2_leaf_reg': 7, 'border_count': 10, 'random_strength': 0.17762021924116494, 'bagging_temperature': 0.9440701490980998}. Best is trial 2 with value: 0.652972744323651.



bestTest = 0.449832505
bestIteration = 273

Training on fold [0/5]

bestTest = 0.6231974712
bestIteration = 621

Training on fold [1/5]

bestTest = 0.6150496754
bestIteration = 621

Training on fold [2/5]

bestTest = 0.6158904971
bestIteration = 621

Training on fold [3/5]

bestTest = 0.6224696903
bestIteration = 621

Training on fold [4/5]


[I 2026-02-14 19:14:03,996] Trial 5 finished with value: 0.6183338676525467 and parameters: {'iterations': 622, 'learning_rate': 0.07050870664563254, 'depth': 3, 'l2_leaf_reg': 6, 'border_count': 87, 'random_strength': 0.366674127298516, 'bagging_temperature': 0.8681270603083875}. Best is trial 2 with value: 0.652972744323651.



bestTest = 0.6150620913
bestIteration = 620

Training on fold [0/5]

bestTest = 0.5777710916
bestIteration = 1479

Training on fold [1/5]

bestTest = 0.5656790309
bestIteration = 1481

Training on fold [2/5]

bestTest = 0.5692064216
bestIteration = 1480

Training on fold [3/5]

bestTest = 0.5851939237
bestIteration = 1478

Training on fold [4/5]


[I 2026-02-14 19:14:23,718] Trial 6 finished with value: 0.5748041500450194 and parameters: {'iterations': 1482, 'learning_rate': 0.02595789180655126, 'depth': 4, 'l2_leaf_reg': 8, 'border_count': 20, 'random_strength': 0.19191717516902745, 'bagging_temperature': 0.4109133728373153}. Best is trial 2 with value: 0.652972744323651.



bestTest = 0.5761980218
bestIteration = 1481

Training on fold [0/5]

bestTest = 0.6693138675
bestIteration = 1311

Training on fold [1/5]

bestTest = 0.6565996069
bestIteration = 1355

Training on fold [2/5]

bestTest = 0.664750941
bestIteration = 1343

Training on fold [3/5]

bestTest = 0.6603479915
bestIteration = 1350

Training on fold [4/5]


[I 2026-02-14 19:14:52,228] Trial 7 finished with value: 0.6620179435831165 and parameters: {'iterations': 1357, 'learning_rate': 0.08602558177305343, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 145, 'random_strength': 0.6890019248339966, 'bagging_temperature': 0.16111621605984794}. Best is trial 7 with value: 0.6620179435831165.



bestTest = 0.6593203567
bestIteration = 1305

Training on fold [0/5]

bestTest = 0.663768432
bestIteration = 1186

Training on fold [1/5]

bestTest = 0.6489055896
bestIteration = 1186

Training on fold [2/5]

bestTest = 0.6570680525
bestIteration = 1189

Training on fold [3/5]

bestTest = 0.6524031458
bestIteration = 1191

Training on fold [4/5]


[I 2026-02-14 19:15:10,077] Trial 8 finished with value: 0.6549260881220647 and parameters: {'iterations': 1193, 'learning_rate': 0.08036404736013536, 'depth': 5, 'l2_leaf_reg': 5, 'border_count': 106, 'random_strength': 0.7525479055557315, 'bagging_temperature': 0.9502290497231303}. Best is trial 7 with value: 0.6620179435831165.



bestTest = 0.6525794821
bestIteration = 1192

Training on fold [0/5]

bestTest = 0.4645606478
bestIteration = 1220

Training on fold [1/5]

bestTest = 0.4671382849
bestIteration = 1220

Training on fold [2/5]

bestTest = 0.4677528923
bestIteration = 1220

Training on fold [3/5]

bestTest = 0.4719907716
bestIteration = 1220

Training on fold [4/5]


[I 2026-02-14 19:15:22,881] Trial 9 finished with value: 0.46765956779278656 and parameters: {'iterations': 1221, 'learning_rate': 0.046691665052421095, 'depth': 1, 'l2_leaf_reg': 8, 'border_count': 135, 'random_strength': 0.737040846572662, 'bagging_temperature': 0.9797625150055067}. Best is trial 7 with value: 0.6620179435831165.



bestTest = 0.4668552423
bestIteration = 1220

Training on fold [0/5]

bestTest = 0.6452823193
bestIteration = 794

Training on fold [1/5]

bestTest = 0.6251183442
bestIteration = 859

Training on fold [2/5]

bestTest = 0.6381698274
bestIteration = 848

Training on fold [3/5]

bestTest = 0.6363501786
bestIteration = 740

Training on fold [4/5]


[I 2026-02-14 19:15:39,794] Trial 10 finished with value: 0.6358979784941996 and parameters: {'iterations': 867, 'learning_rate': 0.09837870291610179, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 47, 'random_strength': 0.5995123583653946, 'bagging_temperature': 0.01045009634474195}. Best is trial 7 with value: 0.6620179435831165.



bestTest = 0.6353937547
bestIteration = 801

Training on fold [0/5]

bestTest = 0.6695545849
bestIteration = 1117

Training on fold [1/5]

bestTest = 0.6553520222
bestIteration = 1071

Training on fold [2/5]

bestTest = 0.6636286401
bestIteration = 1129

Training on fold [3/5]

bestTest = 0.6581823891
bestIteration = 1111

Training on fold [4/5]


[I 2026-02-14 19:16:01,090] Trial 11 finished with value: 0.6609939170600008 and parameters: {'iterations': 1136, 'learning_rate': 0.09787309277909263, 'depth': 6, 'l2_leaf_reg': 5, 'border_count': 149, 'random_strength': 0.7064082332998247, 'bagging_temperature': 0.16361500432646692}. Best is trial 7 with value: 0.6620179435831165.



bestTest = 0.6585987792
bestIteration = 1000

Training on fold [0/5]

bestTest = 0.6698600627
bestIteration = 950

Training on fold [1/5]

bestTest = 0.6568911853
bestIteration = 966

Training on fold [2/5]

bestTest = 0.6661014712
bestIteration = 963

Training on fold [3/5]

bestTest = 0.6611361738
bestIteration = 965

Training on fold [4/5]


[I 2026-02-14 19:16:22,698] Trial 12 finished with value: 0.662692872963749 and parameters: {'iterations': 967, 'learning_rate': 0.09772587219349477, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 150, 'random_strength': 0.5736943060179452, 'bagging_temperature': 0.13163495871344533}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6596760028
bestIteration = 958

Training on fold [0/5]

bestTest = 0.6709383456
bestIteration = 888

Training on fold [1/5]

bestTest = 0.6562179708
bestIteration = 887

Training on fold [2/5]

bestTest = 0.6645355263
bestIteration = 888

Training on fold [3/5]

bestTest = 0.6607846631
bestIteration = 888

Training on fold [4/5]


[I 2026-02-14 19:16:47,168] Trial 13 finished with value: 0.6623349445958429 and parameters: {'iterations': 889, 'learning_rate': 0.08707023481844611, 'depth': 8, 'l2_leaf_reg': 10, 'border_count': 143, 'random_strength': 0.4889439343284035, 'bagging_temperature': 0.28085145305838155}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6592412759
bestIteration = 886

Training on fold [0/5]

bestTest = 0.6581931051
bestIteration = 861

Training on fold [1/5]

bestTest = 0.6401988964
bestIteration = 810

Training on fold [2/5]

bestTest = 0.6482622484
bestIteration = 865

Training on fold [3/5]

bestTest = 0.6491200295
bestIteration = 869

Training on fold [4/5]


[I 2026-02-14 19:17:12,331] Trial 14 finished with value: 0.6483652053255972 and parameters: {'iterations': 870, 'learning_rate': 0.061995992102631896, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 67, 'random_strength': 0.48247003131164956, 'bagging_temperature': 0.3703163686888965}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6462999779
bestIteration = 864

Training on fold [0/5]

bestTest = 0.6685254086
bestIteration = 859

Training on fold [1/5]

bestTest = 0.6548125906
bestIteration = 780

Training on fold [2/5]

bestTest = 0.662381768
bestIteration = 842

Training on fold [3/5]

bestTest = 0.6607442083
bestIteration = 846

Training on fold [4/5]


[I 2026-02-14 19:17:35,021] Trial 15 finished with value: 0.6606776897470005 and parameters: {'iterations': 861, 'learning_rate': 0.08860585077116878, 'depth': 8, 'l2_leaf_reg': 9, 'border_count': 127, 'random_strength': 0.5035544666852528, 'bagging_temperature': 0.23913006148298432}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.657358941
bestIteration = 835

Training on fold [0/5]

bestTest = 0.6585763
bestIteration = 765

Training on fold [1/5]

bestTest = 0.637624163
bestIteration = 763

Training on fold [2/5]

bestTest = 0.6472829267
bestIteration = 760

Training on fold [3/5]

bestTest = 0.6466317181
bestIteration = 762

Training on fold [4/5]


[I 2026-02-14 19:18:10,358] Trial 16 finished with value: 0.6471345555454235 and parameters: {'iterations': 766, 'learning_rate': 0.06244455047459993, 'depth': 10, 'l2_leaf_reg': 10, 'border_count': 63, 'random_strength': 0.5972096817590656, 'bagging_temperature': 0.00018994825415205585}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6456162221
bestIteration = 749

Training on fold [0/5]

bestTest = 0.6668787189
bestIteration = 973

Training on fold [1/5]

bestTest = 0.6531669169
bestIteration = 976

Training on fold [2/5]

bestTest = 0.6609580531
bestIteration = 969

Training on fold [3/5]

bestTest = 0.6587377004
bestIteration = 974

Training on fold [4/5]


[I 2026-02-14 19:18:28,584] Trial 17 finished with value: 0.6592253727809317 and parameters: {'iterations': 977, 'learning_rate': 0.08987634709596434, 'depth': 6, 'l2_leaf_reg': 7, 'border_count': 130, 'random_strength': 0.4668215436467859, 'bagging_temperature': 0.5924993258400937}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6565104638
bestIteration = 970

Training on fold [0/5]

bestTest = 0.669105737
bestIteration = 688

Training on fold [1/5]

bestTest = 0.6554093452
bestIteration = 685

Training on fold [2/5]

bestTest = 0.6623190388
bestIteration = 688

Training on fold [3/5]

bestTest = 0.6613900959
bestIteration = 686

Training on fold [4/5]


[I 2026-02-14 19:18:47,495] Trial 18 finished with value: 0.6613694185409902 and parameters: {'iterations': 689, 'learning_rate': 0.09728271520285267, 'depth': 8, 'l2_leaf_reg': 9, 'border_count': 147, 'random_strength': 0.8171760133960979, 'bagging_temperature': 0.2889643533609537}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6589773375
bestIteration = 657

Training on fold [0/5]

bestTest = 0.6571397565
bestIteration = 973

Training on fold [1/5]

bestTest = 0.6439847482
bestIteration = 974

Training on fold [2/5]

bestTest = 0.6527218903
bestIteration = 963

Training on fold [3/5]

bestTest = 0.6505728055
bestIteration = 973

Training on fold [4/5]


[I 2026-02-14 19:19:04,236] Trial 19 finished with value: 0.6509565997093252 and parameters: {'iterations': 975, 'learning_rate': 0.061823108876943166, 'depth': 5, 'l2_leaf_reg': 8, 'border_count': 125, 'random_strength': 0.5666898808734884, 'bagging_temperature': 0.1093235327365342}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6503973106
bestIteration = 974

Training on fold [0/5]

bestTest = 0.6358274579
bestIteration = 730

Training on fold [1/5]

bestTest = 0.6161547163
bestIteration = 702

Training on fold [2/5]

bestTest = 0.6296272208
bestIteration = 740

Training on fold [3/5]

bestTest = 0.6305702815
bestIteration = 636

Training on fold [4/5]


[I 2026-02-14 19:19:20,642] Trial 20 finished with value: 0.627905804977856 and parameters: {'iterations': 745, 'learning_rate': 0.08352632748105793, 'depth': 8, 'l2_leaf_reg': 6, 'border_count': 37, 'random_strength': 0.276080285625613, 'bagging_temperature': 0.4928335571775482}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6280814903
bestIteration = 557

Training on fold [0/5]

bestTest = 0.6609351934
bestIteration = 513

Training on fold [1/5]

bestTest = 0.6488453059
bestIteration = 513

Training on fold [2/5]

bestTest = 0.6561598394
bestIteration = 504

Training on fold [3/5]

bestTest = 0.6554785713
bestIteration = 513

Training on fold [4/5]


[I 2026-02-14 19:19:32,430] Trial 21 finished with value: 0.6548509949767121 and parameters: {'iterations': 514, 'learning_rate': 0.0876857655463982, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 145, 'random_strength': 0.6571865113211927, 'bagging_temperature': 0.13234646981752646}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6528616914
bestIteration = 513

Training on fold [0/5]

bestTest = 0.6705279495
bestIteration = 1014

Training on fold [1/5]

bestTest = 0.6563282434
bestIteration = 1065

Training on fold [2/5]

bestTest = 0.6650027807
bestIteration = 1044

Training on fold [3/5]

bestTest = 0.6609991785
bestIteration = 1065

Training on fold [4/5]


[I 2026-02-14 19:19:56,299] Trial 22 finished with value: 0.6621678314671195 and parameters: {'iterations': 1066, 'learning_rate': 0.09195641642147082, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 137, 'random_strength': 0.44714428153627955, 'bagging_temperature': 0.25031011902623174}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6582221863
bestIteration = 1057

Training on fold [0/5]

bestTest = 0.6709522105
bestIteration = 1033

Training on fold [1/5]

bestTest = 0.6564731928
bestIteration = 1036

Training on fold [2/5]

bestTest = 0.6637490566
bestIteration = 892

Training on fold [3/5]

bestTest = 0.65948177
bestIteration = 1045

Training on fold [4/5]


[I 2026-02-14 19:20:36,955] Trial 23 finished with value: 0.6619310057931773 and parameters: {'iterations': 1070, 'learning_rate': 0.09293305501399754, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 136, 'random_strength': 0.44636095467274156, 'bagging_temperature': 0.29699879326382317}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6595966864
bestIteration = 946

Training on fold [0/5]

bestTest = 0.6623826599
bestIteration = 961

Training on fold [1/5]

bestTest = 0.6502079216
bestIteration = 962

Training on fold [2/5]

bestTest = 0.6596565121
bestIteration = 962

Training on fold [3/5]

bestTest = 0.6561023708
bestIteration = 962

Training on fold [4/5]


[I 2026-02-14 19:20:55,601] Trial 24 finished with value: 0.6564802775465204 and parameters: {'iterations': 963, 'learning_rate': 0.07950794165343929, 'depth': 6, 'l2_leaf_reg': 10, 'border_count': 121, 'random_strength': 0.3968529731453702, 'bagging_temperature': 0.23439962194779773}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6540928443
bestIteration = 956

Training on fold [0/5]

bestTest = 0.6615005711
bestIteration = 1084

Training on fold [1/5]

bestTest = 0.6499777821
bestIteration = 1073

Training on fold [2/5]

bestTest = 0.6587121072
bestIteration = 1044

Training on fold [3/5]

bestTest = 0.6553533004
bestIteration = 1085

Training on fold [4/5]


[I 2026-02-14 19:21:23,681] Trial 25 finished with value: 0.6555009038184869 and parameters: {'iterations': 1087, 'learning_rate': 0.06837333707588569, 'depth': 8, 'l2_leaf_reg': 8, 'border_count': 97, 'random_strength': 0.537015029774054, 'bagging_temperature': 0.39966570776758703}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.652145297
bestIteration = 1045

Training on fold [0/5]

bestTest = 0.6623808711
bestIteration = 911

Training on fold [1/5]

bestTest = 0.6477863322
bestIteration = 911

Training on fold [2/5]

bestTest = 0.6572482978
bestIteration = 911

Training on fold [3/5]

bestTest = 0.6530243324
bestIteration = 890

Training on fold [4/5]


[I 2026-02-14 19:21:39,350] Trial 26 finished with value: 0.6549072007521969 and parameters: {'iterations': 912, 'learning_rate': 0.09354186395908293, 'depth': 5, 'l2_leaf_reg': 9, 'border_count': 135, 'random_strength': 0.26665698036964613, 'bagging_temperature': 0.0843987814780624}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6542853857
bestIteration = 904

Training on fold [0/5]

bestTest = 0.66009188
bestIteration = 746

Training on fold [1/5]

bestTest = 0.6435620423
bestIteration = 550

Training on fold [2/5]

bestTest = 0.649801706
bestIteration = 849

Training on fold [3/5]

bestTest = 0.6508688994
bestIteration = 785

Training on fold [4/5]


[I 2026-02-14 19:22:16,962] Trial 27 finished with value: 0.6498100643990472 and parameters: {'iterations': 1234, 'learning_rate': 0.09949142858356734, 'depth': 9, 'l2_leaf_reg': 7, 'border_count': 76, 'random_strength': 0.4277878702713372, 'bagging_temperature': 0.5077448415231844}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6464094425
bestIteration = 550

Training on fold [0/5]

bestTest = 0.6664660593
bestIteration = 836

Training on fold [1/5]

bestTest = 0.6541711674
bestIteration = 836

Training on fold [2/5]

bestTest = 0.6614154648
bestIteration = 831

Training on fold [3/5]

bestTest = 0.6572277889
bestIteration = 833

Training on fold [4/5]


[I 2026-02-14 19:22:35,906] Trial 28 finished with value: 0.6589086763503023 and parameters: {'iterations': 837, 'learning_rate': 0.07892945971267891, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 138, 'random_strength': 0.6377131599574501, 'bagging_temperature': 0.32247733736205025}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6554418339
bestIteration = 805

Training on fold [0/5]

bestTest = 0.6420406021
bestIteration = 785

Training on fold [1/5]

bestTest = 0.6313177869
bestIteration = 785

Training on fold [2/5]

bestTest = 0.6368072182
bestIteration = 785

Training on fold [3/5]

bestTest = 0.6354683634
bestIteration = 784

Training on fold [4/5]


[I 2026-02-14 19:22:48,019] Trial 29 finished with value: 0.6361926851993827 and parameters: {'iterations': 786, 'learning_rate': 0.05250126027757621, 'depth': 4, 'l2_leaf_reg': 4, 'border_count': 109, 'random_strength': 0.1053464462208068, 'bagging_temperature': 0.23631054798248652}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6353334369
bestIteration = 785

Training on fold [0/5]

bestTest = 0.6526301865
bestIteration = 1020

Training on fold [1/5]

bestTest = 0.6417197272
bestIteration = 1020

Training on fold [2/5]

bestTest = 0.6489649417
bestIteration = 1017

Training on fold [3/5]

bestTest = 0.6468915553
bestIteration = 1020

Training on fold [4/5]


[I 2026-02-14 19:23:07,290] Trial 30 finished with value: 0.6471910501734597 and parameters: {'iterations': 1021, 'learning_rate': 0.03441099389928046, 'depth': 6, 'l2_leaf_reg': 9, 'border_count': 120, 'random_strength': 0.5307519011643049, 'bagging_temperature': 0.7532962354998036}. Best is trial 12 with value: 0.662692872963749.



bestTest = 0.6457595572
bestIteration = 1019

Training on fold [0/5]

bestTest = 0.6712509821
bestIteration = 1308

Training on fold [1/5]

bestTest = 0.6567394214
bestIteration = 1301

Training on fold [2/5]

bestTest = 0.6650011327
bestIteration = 1311

Training on fold [3/5]

bestTest = 0.6619555898
bestIteration = 1301

Training on fold [4/5]


[I 2026-02-14 19:23:36,916] Trial 31 finished with value: 0.66292845979518 and parameters: {'iterations': 1312, 'learning_rate': 0.08578731679709362, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 148, 'random_strength': 0.6411534370183606, 'bagging_temperature': 0.17658407439582796}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6599064734
bestIteration = 1213

Training on fold [0/5]

bestTest = 0.6720303068
bestIteration = 1149

Training on fold [1/5]

bestTest = 0.6579078411
bestIteration = 1138

Training on fold [2/5]

bestTest = 0.6637708089
bestIteration = 1143

Training on fold [3/5]

bestTest = 0.6609431196
bestIteration = 1149

Training on fold [4/5]


[I 2026-02-14 19:24:07,595] Trial 32 finished with value: 0.6626658844978626 and parameters: {'iterations': 1150, 'learning_rate': 0.09184341697326089, 'depth': 8, 'l2_leaf_reg': 10, 'border_count': 141, 'random_strength': 0.33836065208708177, 'bagging_temperature': 0.07341259458907481}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6588640864
bestIteration = 1132

Training on fold [0/5]

bestTest = 0.6725779226
bestIteration = 1059

Training on fold [1/5]

bestTest = 0.6573425019
bestIteration = 971

Training on fold [2/5]

bestTest = 0.6647700937
bestIteration = 991

Training on fold [3/5]

bestTest = 0.6576247493
bestIteration = 1027

Training on fold [4/5]


[I 2026-02-14 19:24:40,162] Trial 33 finished with value: 0.6621976830326235 and parameters: {'iterations': 1286, 'learning_rate': 0.08338882505979593, 'depth': 8, 'l2_leaf_reg': 1, 'border_count': 150, 'random_strength': 0.34283012577771604, 'bagging_temperature': 0.07418983786360481}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6595103122
bestIteration = 1208

Training on fold [0/5]

bestTest = 0.6694400888
bestIteration = 1128

Training on fold [1/5]

bestTest = 0.6544580479
bestIteration = 967

Training on fold [2/5]

bestTest = 0.6601837371
bestIteration = 1188

Training on fold [3/5]

bestTest = 0.6589300458
bestIteration = 1227

Training on fold [4/5]


[I 2026-02-14 19:25:22,900] Trial 34 finished with value: 0.659499155368908 and parameters: {'iterations': 1310, 'learning_rate': 0.07565050188936248, 'depth': 9, 'l2_leaf_reg': 9, 'border_count': 113, 'random_strength': 0.8439538557277968, 'bagging_temperature': 0.18076758744553417}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6556295821
bestIteration = 933

Training on fold [0/5]

bestTest = 0.6690146932
bestIteration = 1113

Training on fold [1/5]

bestTest = 0.654713796
bestIteration = 1147

Training on fold [2/5]

bestTest = 0.6631088381
bestIteration = 1105

Training on fold [3/5]

bestTest = 0.6608026449
bestIteration = 1143

Training on fold [4/5]


[I 2026-02-14 19:25:50,898] Trial 35 finished with value: 0.6610569449274203 and parameters: {'iterations': 1150, 'learning_rate': 0.09341926793885108, 'depth': 8, 'l2_leaf_reg': 8, 'border_count': 127, 'random_strength': 0.2697212446668063, 'bagging_temperature': 0.040142545666020366}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6583889319
bestIteration = 847

Training on fold [0/5]

bestTest = 0.6709020013
bestIteration = 1293

Training on fold [1/5]

bestTest = 0.656077538
bestIteration = 1398

Training on fold [2/5]

bestTest = 0.662241414
bestIteration = 1406

Training on fold [3/5]

bestTest = 0.6608191954
bestIteration = 1177

Training on fold [4/5]


[I 2026-02-14 19:27:31,875] Trial 36 finished with value: 0.661489740959009 and parameters: {'iterations': 1442, 'learning_rate': 0.0657795453808609, 'depth': 10, 'l2_leaf_reg': 10, 'border_count': 141, 'random_strength': 0.5792985504656055, 'bagging_temperature': 0.20046300310162943}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6581118003
bestIteration = 1423

Training on fold [0/5]

bestTest = 0.6641319566
bestIteration = 1132

Training on fold [1/5]

bestTest = 0.6534171206
bestIteration = 1066

Training on fold [2/5]

bestTest = 0.660869218
bestIteration = 1129

Training on fold [3/5]

bestTest = 0.6555711579
bestIteration = 1123

Training on fold [4/5]


[I 2026-02-14 19:28:02,320] Trial 37 finished with value: 0.6572566436419471 and parameters: {'iterations': 1135, 'learning_rate': 0.07461466881506873, 'depth': 8, 'l2_leaf_reg': 9, 'border_count': 102, 'random_strength': 0.39360472618240344, 'bagging_temperature': 0.06997586635289726}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6523976168
bestIteration = 1134

Training on fold [0/5]

bestTest = 0.6549797983
bestIteration = 1283

Training on fold [1/5]

bestTest = 0.6422408665
bestIteration = 1283

Training on fold [2/5]

bestTest = 0.6466963581
bestIteration = 1282

Training on fold [3/5]

bestTest = 0.6460818414
bestIteration = 1283

Training on fold [4/5]


[I 2026-02-14 19:28:48,813] Trial 38 finished with value: 0.6472310967158238 and parameters: {'iterations': 1284, 'learning_rate': 0.013045489475512893, 'depth': 9, 'l2_leaf_reg': 3, 'border_count': 116, 'random_strength': 0.29808948584376654, 'bagging_temperature': 0.12689711790348512}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6461700468
bestIteration = 1282

Training on fold [0/5]

bestTest = 0.6691344326
bestIteration = 1392

Training on fold [1/5]

bestTest = 0.6550488721
bestIteration = 1399

Training on fold [2/5]

bestTest = 0.6622989911
bestIteration = 1399

Training on fold [3/5]

bestTest = 0.6583954223
bestIteration = 1399

Training on fold [4/5]


[I 2026-02-14 19:29:15,066] Trial 39 finished with value: 0.6605876357730252 and parameters: {'iterations': 1400, 'learning_rate': 0.08346394081795931, 'depth': 6, 'l2_leaf_reg': 7, 'border_count': 131, 'random_strength': 0.6380310530572143, 'bagging_temperature': 0.4701046746089047}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6583257359
bestIteration = 1381

Training on fold [0/5]

bestTest = 0.6434744817
bestIteration = 921

Training on fold [1/5]

bestTest = 0.6343499165
bestIteration = 920

Training on fold [2/5]

bestTest = 0.6348235315
bestIteration = 921

Training on fold [3/5]

bestTest = 0.6383432967
bestIteration = 921

Training on fold [4/5]


[I 2026-02-14 19:29:34,700] Trial 40 finished with value: 0.6371942940462638 and parameters: {'iterations': 922, 'learning_rate': 0.024649912283989578, 'depth': 7, 'l2_leaf_reg': 10, 'border_count': 93, 'random_strength': 0.9499002637339431, 'bagging_temperature': 0.33708376663997697}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6349812709
bestIteration = 921

Training on fold [0/5]

bestTest = 0.6724519325
bestIteration = 918

Training on fold [1/5]

bestTest = 0.657374232
bestIteration = 897

Training on fold [2/5]

bestTest = 0.6638533966
bestIteration = 953

Training on fold [3/5]

bestTest = 0.6593693028
bestIteration = 987

Training on fold [4/5]


[I 2026-02-14 19:30:11,723] Trial 41 finished with value: 0.6624906145637091 and parameters: {'iterations': 1325, 'learning_rate': 0.08400176371479963, 'depth': 8, 'l2_leaf_reg': 2, 'border_count': 147, 'random_strength': 0.34089819246937253, 'bagging_temperature': 0.0692143807467141}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6605034547
bestIteration = 790

Training on fold [0/5]

bestTest = 0.6715295963
bestIteration = 1296

Training on fold [1/5]

bestTest = 0.6560953746
bestIteration = 1087

Training on fold [2/5]

bestTest = 0.6640319398
bestIteration = 965

Training on fold [3/5]

bestTest = 0.6589569112
bestIteration = 1030

Training on fold [4/5]


[I 2026-02-14 19:30:42,728] Trial 42 finished with value: 0.6620020642143478 and parameters: {'iterations': 1337, 'learning_rate': 0.08744609643808514, 'depth': 7, 'l2_leaf_reg': 2, 'border_count': 142, 'random_strength': 0.21447245770293627, 'bagging_temperature': 0.043254580879016236}. Best is trial 31 with value: 0.66292845979518.



bestTest = 0.6605746774
bestIteration = 1190

Training on fold [0/5]

bestTest = 0.6730292849
bestIteration = 884

Training on fold [1/5]

bestTest = 0.6593345667
bestIteration = 921

Training on fold [2/5]

bestTest = 0.6655408321
bestIteration = 1034

Training on fold [3/5]

bestTest = 0.6602333026
bestIteration = 799

Training on fold [4/5]


[I 2026-02-14 19:31:23,817] Trial 43 finished with value: 0.6636717394267887 and parameters: {'iterations': 1485, 'learning_rate': 0.09566208338756467, 'depth': 8, 'l2_leaf_reg': 4, 'border_count': 149, 'random_strength': 0.3221392635228121, 'bagging_temperature': 0.16171642890213989}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6609619792
bestIteration = 947

Training on fold [0/5]

bestTest = 0.670755191
bestIteration = 613

Training on fold [1/5]

bestTest = 0.6565449237
bestIteration = 783

Training on fold [2/5]

bestTest = 0.6642340211
bestIteration = 756

Training on fold [3/5]

bestTest = 0.661421127
bestIteration = 709

Training on fold [4/5]


[I 2026-02-14 19:32:19,254] Trial 44 finished with value: 0.6622627756227637 and parameters: {'iterations': 1455, 'learning_rate': 0.09652330438680544, 'depth': 9, 'l2_leaf_reg': 3, 'border_count': 150, 'random_strength': 0.33259158291965263, 'bagging_temperature': 0.1395109221978586}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6599009085
bestIteration = 554

Training on fold [0/5]

bestTest = 0.6709672872
bestIteration = 938

Training on fold [1/5]

bestTest = 0.6571909879
bestIteration = 1346

Training on fold [2/5]

bestTest = 0.6653473968
bestIteration = 1067

Training on fold [3/5]

bestTest = 0.6603746307
bestIteration = 790

Training on fold [4/5]


[I 2026-02-14 19:32:51,407] Trial 45 finished with value: 0.6625705410274818 and parameters: {'iterations': 1482, 'learning_rate': 0.09978229600499584, 'depth': 7, 'l2_leaf_reg': 4, 'border_count': 150, 'random_strength': 0.12271876453671643, 'bagging_temperature': 0.18585170978314647}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6600893586
bestIteration = 1005

Training on fold [0/5]

bestTest = 0.6678905332
bestIteration = 1277

Training on fold [1/5]

bestTest = 0.6561197768
bestIteration = 1352

Training on fold [2/5]

bestTest = 0.6649772014
bestIteration = 1479

Training on fold [3/5]

bestTest = 0.6574524949
bestIteration = 1194

Training on fold [4/5]


[I 2026-02-14 19:33:20,331] Trial 46 finished with value: 0.660730036023816 and parameters: {'iterations': 1498, 'learning_rate': 0.09473900656225046, 'depth': 6, 'l2_leaf_reg': 4, 'border_count': 132, 'random_strength': 0.1393643645704797, 'bagging_temperature': 0.20342411898455565}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6578310458
bestIteration = 1394

Training on fold [0/5]

bestTest = 0.6304942861
bestIteration = 1397

Training on fold [1/5]

bestTest = 0.6226371279
bestIteration = 1391

Training on fold [2/5]

bestTest = 0.6279589086
bestIteration = 1387

Training on fold [3/5]

bestTest = 0.6275554635
bestIteration = 1388

Training on fold [4/5]


[I 2026-02-14 19:33:36,571] Trial 47 finished with value: 0.6276267782860694 and parameters: {'iterations': 1398, 'learning_rate': 0.09882091162480786, 'depth': 2, 'l2_leaf_reg': 4, 'border_count': 140, 'random_strength': 0.2040653955304011, 'bagging_temperature': 0.18865630152244967}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6295818849
bestIteration = 1397

Training on fold [0/5]

bestTest = 0.6686072891
bestIteration = 1193

Training on fold [1/5]

bestTest = 0.652739669
bestIteration = 1169

Training on fold [2/5]

bestTest = 0.6628912971
bestIteration = 1213

Training on fold [3/5]

bestTest = 0.6594659097
bestIteration = 1239

Training on fold [4/5]


[I 2026-02-14 19:34:04,071] Trial 48 finished with value: 0.6599681164376175 and parameters: {'iterations': 1243, 'learning_rate': 0.09999100412807026, 'depth': 7, 'l2_leaf_reg': 5, 'border_count': 123, 'random_strength': 0.7662050201728497, 'bagging_temperature': 0.10607019813500934}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6571424974
bestIteration = 780

Training on fold [0/5]

bestTest = 0.6698890952
bestIteration = 1122

Training on fold [1/5]

bestTest = 0.6557355416
bestIteration = 1358

Training on fold [2/5]

bestTest = 0.6649958232
bestIteration = 1014

Training on fold [3/5]

bestTest = 0.6581689965
bestIteration = 1204

Training on fold [4/5]


[I 2026-02-14 19:34:36,557] Trial 49 finished with value: 0.6612281408973046 and parameters: {'iterations': 1440, 'learning_rate': 0.09298784720422713, 'depth': 7, 'l2_leaf_reg': 3, 'border_count': 140, 'random_strength': 0.2324883055329317, 'bagging_temperature': 0.8346897095108177}. Best is trial 43 with value: 0.6636717394267887.



bestTest = 0.6582229982
bestIteration = 1101

Best Parameters: {'iterations': 1485, 'learning_rate': 0.09566208338756467, 'depth': 8, 'l2_leaf_reg': 4, 'border_count': 149, 'random_strength': 0.3221392635228121, 'bagging_temperature': 0.16171642890213989}


In [ ]:
# validating the tuned CatBoostRegressor
best_params = study.best_params

tuned_CBR = CatBoostRegressor(**best_params)
tuned_CBR.fit(X_train, y_train)
y_pred = tuned_CBR.predict(X_val)

catboost_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

catboost_results_df = pd.DataFrame([catboost_results])

In [21]:
catboost_results_df

MSE       MAE      RMSE  R2 Score
0  0.045235  0.159138  0.212686  0.644328

### 2. XGBRegressor

In [23]:
# objective function for tuning XGBRegressor
def objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.2),
        'max_depth': trial.suggest_int('max_depth', 1, 10),
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 10),
        'subsample': trial.suggest_float('subsample',0.5, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0),
        'gamma': trial.suggest_float('gamma', 0, 0.5),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 1.0),
        'reg_lambda': trial.suggest_float('reg_lambda', 1.0, 2.5),
        'random_state': SEED
    }

    model = XGBRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=FOLDS, scoring='r2').mean()

    return score

In [24]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-14 19:37:20,856] A new study created in memory with name: no-name-bbdd3908-1049-4acb-be98-8b827e190b67
C:\Users\ansar\Desktop\Workspace\Personal\MLOPs\House Price Prediction\venv\Lib\site-packages\xgboost\training.py:200: UserWarning: [19:37:20] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "random_seed" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
C:\Users\ansar\Desktop\Workspace\Personal\MLOPs\House Price Prediction\venv\Lib\site-packages\xgboost\training.py:200: UserWarning: [19:37:21] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "random_seed" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)
C:\Users\ansar\Desktop\Workspace\Personal\MLOPs\House Price Prediction\venv\Lib\site-packages\xgboost\training.py:200: UserWarning: [19:37:22] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:782: 
Parameters: { "random_seed" } are not used.

  bst.update(dtrain, ite

Best Parameters: {'n_estimators': 535, 'learning_rate': 0.057582817878155085, 'max_depth': 10, 'min_child_weight': 10, 'subsample': 0.7124624923071757, 'colsample_bytree': 0.7561466714445289, 'gamma': 0.0433184458126587, 'reg_alpha': 0.934442472471591, 'reg_lambda': 2.07724971668887}


In [25]:
# validating the tuned XGBRegressor
best_params = study.best_params

tuned_XGB = XGBRegressor(**best_params)
tuned_XGB.fit(X_train, y_train)
y_pred = tuned_XGB.predict(X_val)

xgb_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

xgb_results_df = pd.DataFrame([xgb_results])

In [26]:
xgb_results_df

MSE       MAE      RMSE  R2 Score
0  0.043301  0.155576  0.208089  0.659535

### 3. LGBMRegressor

In [8]:
# objective function for LGBMRegressor

def objective(trial):
    params = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.2),
        'max_depth': trial.suggest_int('max_depth', 3, 9),
        'num_leaves': trial.suggest_int('num_leaves', 31, 200),
        'min_child_samples': trial.suggest_int('min_child_samples', 20, 200),
        'subsample': trial.suggest_float('subsample', 0.8, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.8, 1.0),
        'reg_alpha': trial.suggest_float('reg_alpha', 0, 1.0),
        'reg_lambda': trial.suggest_float('reg_lambda', 0, 1.0),
        'random_state': SEED,
        'verbose':-1
    }

    model = LGBMRegressor(**params)
    score = cross_val_score(model, X_train, y_train, cv=FOLDS, scoring='r2').mean()

    return score

In [9]:
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

print(f'Best Parameters: {study.best_params}')

[I 2026-02-14 20:07:05,525] A new study created in memory with name: no-name-ffde3516-3541-4fcd-85f3-55b7b241bd54
[I 2026-02-14 20:07:10,823] Trial 0 finished with value: 0.6671900591722856 and parameters: {'n_estimators': 921, 'learning_rate': 0.08289229517955117, 'max_depth': 7, 'num_leaves': 185, 'min_child_samples': 29, 'subsample': 0.9071106279880145, 'colsample_bytree': 0.8262395832621486, 'reg_alpha': 0.6294709479551285, 'reg_lambda': 0.8995915270683216}. Best is trial 0 with value: 0.6671900591722856.
[I 2026-02-14 20:07:12,626] Trial 1 finished with value: 0.6608516078554858 and parameters: {'n_estimators': 616, 'learning_rate': 0.0617964460707792, 'max_depth': 5, 'num_leaves': 98, 'min_child_samples': 133, 'subsample': 0.9408333994198707, 'colsample_bytree': 0.851449526148705, 'reg_alpha': 0.5966186211772637, 'reg_lambda': 0.18656031323775435}. Best is trial 0 with value: 0.6671900591722856.
[I 2026-02-14 20:07:14,861] Trial 2 finished with value: 0.6644826147368906 and param

Best Parameters: {'n_estimators': 214, 'learning_rate': 0.10979784611692994, 'max_depth': 9, 'num_leaves': 75, 'min_child_samples': 33, 'subsample': 0.939496220559003, 'colsample_bytree': 0.8775803119044167, 'reg_alpha': 0.8615502636228614, 'reg_lambda': 0.8363793385956172}


In [10]:
# validating the tuned LGBMRegressor
best_params = study.best_params

tuned_LGBM = LGBMRegressor(**best_params)
tuned_LGBM.fit(X_train, y_train)
y_pred = tuned_LGBM.predict(X_val)

lgbm_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

lgbm_results_df = pd.DataFrame([lgbm_results])

In [11]:
lgbm_results_df

MSE       MAE      RMSE  R2 Score
0  0.043555  0.155852  0.208699  0.657535

### 4. Ensemble of tuned models

In [9]:
# ensemble of tuned models
ensemble = StackingRegressor(
    estimators=[
        ('catboost', tuned_CBR),
        ('lgbm', tuned_LGBM),
        ('xgb', tuned_XGB)
], final_estimator=Ridge(), cv=FOLDS)

In [10]:
# training the ensemble on training dataset
ensemble.fit(X_train, y_train)
y_pred = ensemble.predict(X_val)

ensemble_results = {
    'MSE': mean_squared_error(y_val, y_pred),
    'MAE': mean_absolute_error(y_val, y_pred),
    'RMSE': root_mean_squared_error(y_val, y_pred),
    'R2 Score': r2_score(y_val, y_pred)
}

ensemble_results_df = pd.DataFrame([ensemble_results])

0:	learn: 0.3427820	total: 151ms	remaining: 3m 44s
1:	learn: 0.3303523	total: 158ms	remaining: 1m 56s
2:	learn: 0.3196038	total: 163ms	remaining: 1m 20s
3:	learn: 0.3105680	total: 168ms	remaining: 1m 2s
4:	learn: 0.3023515	total: 174ms	remaining: 51.4s
5:	learn: 0.2950244	total: 179ms	remaining: 44.1s
6:	learn: 0.2883986	total: 184ms	remaining: 38.8s
7:	learn: 0.2833888	total: 189ms	remaining: 34.9s
8:	learn: 0.2785328	total: 194ms	remaining: 31.8s
9:	learn: 0.2745195	total: 199ms	remaining: 29.3s
10:	learn: 0.2707785	total: 204ms	remaining: 27.3s
11:	learn: 0.2672414	total: 209ms	remaining: 25.6s
12:	learn: 0.2642183	total: 214ms	remaining: 24.2s
13:	learn: 0.2619823	total: 219ms	remaining: 23s
14:	learn: 0.2594563	total: 224ms	remaining: 21.9s
15:	learn: 0.2574520	total: 229ms	remaining: 21s
16:	learn: 0.2557050	total: 233ms	remaining: 20.1s
17:	learn: 0.2540944	total: 238ms	remaining: 19.4s
18:	learn: 0.2521885	total: 244ms	remaining: 18.8s
19:	learn: 0.2504778	total: 249ms	remainin

In [11]:
ensemble_results_df

MSE       MAE      RMSE  R2 Score
0  0.04313  0.155094  0.207677  0.660881

The ensemble model combining CatBoost, XGBoost, and LightGBM achieved the best overall performance, with an MSE of 0.04313, MAE of 0.15509, RMSE of 0.20768, and an R² score of 0.66088.
